Placeholder for Blog for Service CIDR (#49989)
kubernetes/website: content/en/docs/tasks/network/reconfigure-default-service-ip-ranges.md

title: Kubernetes Default Service CIDR Reconfiguration
content_type: task
---

<!-- overview -->
{{< feature-state feature_gate_name="MultiCIDRServiceAllocator" >}}

This document shares how to reconfigure the default Service IP range(s) assigned
to a cluster.

## {{% heading "prerequisites" %}}

{{< include "task-tutorial-prereqs.md" >}}

{{< version-check >}}

<!-- steps -->

## Kubernetes Default Service CIDR Reconfiguration

This document explains how to manage the Service IP address range within a
Kubernetes cluster, that directly influences the cluster's supported IP families
for Services.

The IP families available for Service ClusterIPs are determined by the
`--service-cluster-ip-range` flag to kube-apiserver. For a better understanding of Service IP address allocation, refer to the
[Services IP address allocation tracking](https://kubernetes.io/docs/reference/networking/virtual-ips/#ip-address-objects) documentation.

Since Kubernetes 1.33, the `ServiceCIDR` object named `kubernetes` reflects the
Service IP families configured for the cluster. The `kubernetes` `ServiceCIDR`
object is created by the first kube-apiserver instance that starts, based on its

> configured `--service-cluster-ip-range` flag. To ensure consistent cluster behavior, all kube-apiserver instances must be configured with the same `--service-cluster-ip-range` values that match the default kubernetes ServiceCIDR object.

Reviewer: ```suggestion
configured `--service-cluster-ip-range` flag. To ensure consistent cluster behavior, all kube-apiserver instances must be configured with the same `--service-cluster-ip-range` values, which must match the default kubernetes ServiceCIDR object.
```

    kube-proxy, the CNI or network plugin, service mesh implementations, and DNS
    services.

* **Dual-to-single conversion preserving the primary service CIDR:** This
    involves removing the secondary IP family from a dual-stack cluster,
    reverting to a single IP family while retaining the original primary IP
    family. In addition to the reconfiguration of the components to match the
    new IP family, you might need to address Services that were explicitly
    configured to use the removed IP family.

* **Anything that results in changing the primary service CIDR:** Completely
  replacing the default ServiceCIDR is a complex operation. If the new
  ServiceCIDR does not overlap with the existing one, [it will require
  renumbering all existing Services and changing the `kubernetes.default`
  service](#Illustrative Reconfiguration Steps). The case where the primary IP
  family also changes is even more complicated, and may require to change
  multiple cluster components (kubelet, network plugins, etc.) to match the new
  primary IP family. Due to the complexity of this change, this operation is

>   unsupported.

Author: we allow to reconfigure the service cidr meanwhile is the same IP family,

>   unsupported.

Author: I will remove the sentence, it is doable so no need to make support estatements, just that is complex and you need a lot of stuff should be enough

### Manual Operations for Replacing the Default Service CIDR

> Reconfiguring the default Service CIDR, necessitates manual steps performed by

Reviewer: ```suggestion
Reconfiguring the default Service CIDR necessitates manual steps performed by
```

---
reviewers:
- thockin
- dwinship
min-kubernetes-server-version: v1.33
title: Kubernetes Default Service CIDR Reconfiguration
content_type: task
---

<!-- overview -->
{{< feature-state feature_gate_name="MultiCIDRServiceAllocator" >}}

This document shares how to reconfigure the default Service IP range(s) assigned
to a cluster.

## {{% heading "prerequisites" %}}

{{< include "task-tutorial-prereqs.md" >}}

{{< version-check >}}

<!-- steps -->

## Kubernetes Default Service CIDR Reconfiguration

This document details how to reconfigure the existing Service IP address range

> within a Kubernetes cluster, which in turn affects the cluster's networking
> type, as outlined in the [Kubernetes Networking
> Concepts](https://kubernetes.io/docs/concepts/cluster-administration/
> networking/).

Reviewer: Is this more AI-generated text? The linked page does define "networking type" that way, but nobody ever says that, and I think that page should be rewritten. "Networking type" sounds like it means "what network plugin you use" or something like that.

```suggestion
within a Kubernetes cluster, which may affect the cluster's supported IP families.
```

(The "may" there is also important: you could change the primary ServiceCIDR without changing its IP family.)

Author: the url link is broken because I wrongly wrapped the line and broke it https://kubernetes.io/docs/concepts/cluster-administration/networking/

Author: let me rephrase, indeed the AI isnot helping, is making the text for confusing instead of helping me to make it simpler

Author: removed most of the AI text and simplified with your suggestions, and focus on services ip ranges only

> The networking type of a Kubernetes cluster (IPv4 only, IPv6 only, or
> dual-stack) is largely determined by how the Services ip ranges are configured
> in the kube-apiserver (using the `--service-cluster-ip-range` flag). This
> configuration dictates the IP family (or families) from which ClusterIPs are
> allocated to Services. **The primary IP family configured in the kube-apiserver
> serves as the default IP family for allocating the Services ClusterIPs. It is
> important to note that Kubernetes can only allocate Service IPs from the IP
> families defined in the `--service-cluster-ip-range` flag.** The primary IP
> family of the cluster is typically represented by the IP family of the
> `kubernetes.default` Service ClusterIP.

Reviewer: ```suggestion
The IP families available for Service ClusterIPs are determined by the
`--service-cluster-ip-range` flag to kube-apiserver; if that contains a single
CIDR then you can only create ServiceCIDR objects with CIDRs of the same
IP family. If it contains dual-stack CIDRs then you can create ServiceCIDRs
with CIDRs of both IP families.
```

> With the `MultiCIDRServiceAllocator` feature gate enabled, the default
> ServiceCIDR configuration is represented by a `ServiceCIDR` object named
> `kubernetes`. This object reflects the Service IP families configured for the
> cluster. The `kubernetes` `ServiceCIDR` object is created by the first
> kube-apiserver instance that starts after the feature gate is enabled, based on
> its configured `--service-cluster-ip-range` flag. Subsequent kube-apiserver
> instances will not attempt to update or delete this object, except for adding a
> new CIDR from a different IP family. This capability is essential for
> transitioning the cluster from a single-stack (IPv4-only or IPv6-only) to a
> dual-stack (IPv4/IPv6 or IPv6/IPv4) configuration

Reviewer: ```suggestion
kube-apiserver will create a `ServiceCIDR` object named `kubernetes` to
represent the CIDRs configured via `--service-cluster-ip-range`. This object will
remain unchanged after creation, unless you restart a formerly-single-stack
kube-apiserver with a dual-stack `--service-cluster-ip-range`, in which case it
will add the second CIDR to the `kubernetes` `ServiceCIDR`.
```

1. Don't mention the feature gate any more.
2. Don't talk about multiple apiservers, etc; that's not relevant to users.

Author: > Don't mention the feature gate any more.

yeah, absolutely

> Don't talk about multiple apiservers, etc; that's not relevant to users

his is the documentation that people that develop installers or deploy their own clusters use, shouldn't they be aware of these details?

Reviewer: I guess we need to document it enough for that case, but it seemed like the original text was documenting too many minor implementation details...

If any kube-apiserver in the cluster does not have a configuration matching the
default `kubernetes` `ServiceCIDR` (i.e., the `--service-cluster-ip-range` flag
has different values), it will warn users and operators to rectify the
situation. This consistency mechanism prevents the undefined behavior observed
in earlier versions where multiple apiservers running with different configured
IP ranges or families could lead to issues such as Service IP leaks, incorrect

> IP assignments, or the creation of duplicate IPs.

Reviewer: I'm not sure we need to say this.

Author: I'm just trying to explain the reason why this work the way that it works, to avoid later questions on github issues ... this way any question can be answered by referring to the docs

{{< note >}} It's important to note that any change beyond simply extending the
Service CIDRs is a complex operation requiring careful planning and manual
intervention across various Kubernetes components. Consider the potential impact

> and complexity before proceeding. {{</ note >}}

Reviewer: I don't think this belongs in this section, since you haven't really said anything else about (user) modification of ServiceCIDRs yet.

### Kubernetes Service CIDR Reconfiguration Categories

We can categorize Service CIDR reconfiguration into the following scenarios:

> * **Extending the existing Service CIDRs:** This can be done dynamically adding
>     new ServiceCIDR objects without the need of reconfiguration of the
>     kube-apiserver, please refer to the dedicated Kubernetes documentation on
>     [Extending Service IP
>     Ranges](https://kubernetes.io/docs/tasks/network/extend-service-ip-ranges/).

Reviewer: ```suggestion
* **Extending the existing Service CIDRs:** This can be done dynamically by adding
    new ServiceCIDR objects without the need of reconfiguration of the
    kube-apiserver. Please refer to the dedicated Kubernetes documentation on
    [Extending Service IP
    Ranges](https://kubernetes.io/docs/tasks/network/extend-service-ip-ranges/).
```

* **Single-to-dual-stack conversion preserving the primary service CIDR:** This
    involves introducing a secondary IP family (IPv6 to an IPv4-only cluster, or
    IPv4 to an IPv6-only cluster) while keeping the original IP family as
    primary. This requires an update to the kube-apiserver configuration and a
    corresponding modification of various cluster components that need to handle
    this additional IP family. These components include, but are not limited to,
    kube-proxy, the CNI or network plugin, service mesh implementations, and DNS
    services.

* **Dual-to-single conversion preserving the primary service CIDR:** This
    involves removing the secondary IP family from a dual-stack cluster,
    reverting to a single IP family while retaining the original primary IP
    family. In addition to the reconfiguration of the components to match the
    new IP family, you might need to address Services that were explicitly
    configured to use the removed IP family.

> * **Anything that results in changing the primary service CIDR:** This includes
>     scenarios like swapping the primary IP family in a dual-stack cluster or
>     completely switching from IPv4-only to IPv6-only (or vice versa). This type

Reviewer: ```suggestion
* **Any reconfiguration that changes the primary service CIDR:** This includes
    scenarios like swapping the primary IP family in a dual-stack cluster,
    completely switching from IPv4-only to IPv6-only (or vice versa), or simply
    moving from one CIDR range to a different CIDR range within the same IP
    family. This type
```

There _are_ people who want to change the default ServiceCIDR without changing IP families. Generally "oops I didn't realize the default Service CIDR overlaps a CIDR range I'm using outside the cluster". I think this runs into exactly the same problems as the changing-primary-IP-family case (in particular, you need to move kubernetes.default), so it should be easy to talk about them together.

Author: good point, this is about changing the primary service cidr, and there can be two cases, same IP family per the case you explain, or changing the primary ip family that has more consequences

>     of reconfiguration is undefined and can have catastrophic consequences for

Reviewer: "undefined" doesn't feel right. I'd say it's "not supported", and also that there's no set of instructions for doing it that is guaranteed to work.

Well... except that then you explain how to do it in the next section...

Author: yep

2.  **Reconfiguring** the network components: This is a critical step and the
    specific procedure depends on the different networking components in use. It
    might involve updating configuration files, restarting agent pods, or
    updating the components to manage the new Service CIDR(s) and the desired IP
    family configuration for Pods. Typical components can be the implementation
    of Kubernetes Services, such as kube-proxy, and the configured networking
    plugin, and potentially other networking components like service mesh
    controllers and DNS servers, to ensure they can correctly handle traffic and
    perform service discovery with the new IP family configuration.
3.  **Managing existing Services:** Services with IPs from the old CIDR need to
    be addressed. Options include recreation (leading to downtime and new IP
    assignments) or potentially more complex reconfiguration strategies.
4.  **Recreating internal Kubernetes services:** The `kubernetes.default`
    service must be deleted and recreated to obtain an IP address from the new
    Service CIDR if the primary IP family is changed or replaced by a different

>     network.

Reviewer: in some cases you need to delete the `kubernetes` `ServiceCIDR` too, right?

Author: > obtain an IP address from the new  Service CIDR if the primary IP family is changed or replaced by a different
    network.

this assumes that you have deleted the ServiceCIDR and the primary IP of that Service CIDR does not match the one assigned to the kubernetes.default Service ... I feel a bit frustrated because I'm not able to explain clearly these things 😓
